src/CloudService/custom/New-AzCloudServiceDiagnosticsExtension.ps1

```diff
@@ -0,0 +1,117 @@
+    [string] $StorageAccountName,
+
+    [Parameter(HelpMessage="Storage Account Key.", Mandatory)]
+    [string] $StorageAccountKey,
+
+    [Parameter(HelpMessage="Specifies the version of the extension.")]
+    [string] $TypeHandlerVersion,
+
+    [string[]] $RolesAppliedTo,
+
+    [Boolean] $AutoUpgradeMinorVersion
+  )
+
+  process {
+    $publisher = "Microsoft.Azure.Diagnostics"
+    $extensionType = "PaaSDiagnostics"
+
+    if (!(Test-Path $DiagnosticsConfigurationPath -PathType Leaf))
+    {
+        throw ("DiagnosticsConfigurationPath does not exits: " + $DiagnosticsConfigurationPath)
+    }
+
+    if (-not $Subscription)
+    {
+        $context = Get-AzContext
+        $Subscription = $context.Subscription.Id
+    }
+
+    [xml]$diagnosticsConfigurationXml = Get-Content $DiagnosticsConfigurationPath
+
+    $storageAccount = $diagnosticsConfigurationXml.PublicConfig.ChildNodes | Where-Object { $_.Name -eq 'StorageAccount' }
+    if ($storageAccount)
+    {
+        Write-Host "Using StorageAccount information defined in diagnostics configuration file."
+    }
+    else
+    {
+        $storageAccount = $diagnosticsConfigurationXml.CreateElement('StorageAccount', $diagnosticsConfigurationXml.PublicConfig.NamespaceURI)
+        $storageAccount.InnerText = $StorageAccountName
+        $storageAccount = $diagnosticsConfigurationXml.PublicConfig.AppendChild($storageAccount)
+    }
+
+    $metrics = $diagnosticsConfigurationXml.PublicConfig.WadCfg.DiagnosticMonitorConfiguration.ChildNodes | Where-Object { $_.Name -eq 'Metrics' }
+    if ($metrics)
+    {
+        Write-Host "Using Metrics information defined in diagnostics configuration file."
+    }
+    else
+    {
+        $metrics = $diagnosticsConfigurationXml.CreateElement('Metrics', $diagnosticsConfigurationXml.PublicConfig.NamespaceURI)
+        $resourceId = '/subscriptions/' + $Subscription + '/resourceGroups/'+ $ResourceGroupName + '/providers/Microsoft.Compute/cloudservices/' + $CloudServiceName
+        $metrics.SetAttribute('resourceId', $resourceId)
+        $metrics = $diagnosticsConfigurationXml.PublicConfig.WadCfg.DiagnosticMonitorConfiguration.AppendChild($metrics)
+    }
+
+    $setting = $diagnosticsConfigurationXml.PublicConfig.OuterXml
+
+    $privateConfig = $diagnosticsConfigurationXml.ChildNodes | Where-Object { $_.Name -eq 'PrivateConfig' }
+    if ($privateConfig)
+    {
+        $protectedSetting = $privateConfig.OuterXml
+        Write-Host "Using PrivateConfig information defined in diagnostics configuration file."
+    }
+    else
+    {
+        $storageEndpoint = "https://core.windows.net"
+        $protectedSetting = '<?xml version="1.0" encoding="UTF-8"?><PrivateConfig xmlns="http://schemas.microsoft.com/ServiceHosting/2010/10/DiagnosticsConfiguration"><StorageAccount endpoint="' + $storageEndpoint + '" key="' + $StorageAccountKey + '" name="' + $storageAccountName + '"/></PrivateConfig>'
```

Comment:
It is better to use "${variable}" in the string.

Reply:
There will be many escape char for `"` in xml.